internal/telemetry/client.go

```diff
@@ -131,87 +138,91 @@ type Client struct {
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	if c.Disabled || !internal.BoolEnv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", true) {
+	if !internal.BoolEnv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", true) || c.Disabled {
 		return
 	}
 	if c.started {
 		return
 	}
-	c.debug = internal.BoolEnv("DD_INSTRUMENTATION_TELEMETRY_DEBUG", c.debug)
-
 	c.started = true
 
 	// XXX: Should we let metrics persist between starting and stopping?
 	c.metrics = make(map[string]*metric)
+	c.applyDefaultOps()
 
 	payload := &AppStarted{
-		Integrations:  append([]Integration{}, integrations...),
 		Configuration: append([]Configuration{}, configuration...),
-	}
-	deps, ok := debug.ReadBuildInfo()
-	if ok {
-		for _, dep := range deps.Deps {
-			payload.Dependencies = append(payload.Dependencies,
-				Dependency{
-					Name:    dep.Path,
-					Version: dep.Version,
-				},
-			)
-		}
+		Products: Products{
+			AppSec: ProductDetails{
+				Version: version.Tag,
+				Enabled: appsec.Enabled(),
+			},
+		},
 	}
 
-	// configEnvFallback returns the value of environment variable with the
-	// given key if def == ""
-	configEnvFallback := func(key, def string) string {
-		if def != "" {
-			return def
-		}
-		return os.Getenv(key)
-	}
-	c.Service = configEnvFallback("DD_SERVICE", c.Service)
-	if len(c.Service) == 0 {
-		if name := globalconfig.ServiceName(); len(name) != 0 {
-			c.Service = name
-		} else {
-			// I think service *has* to be something?
-			c.Service = "unnamed-go-service"
+	appStarted := c.newRequest(RequestTypeAppStarted)
+	appStarted.Body.Payload = payload
+	c.scheduleSubmit(appStarted)
+
+	if internal.BoolEnv("DD_TELEMETRY_DEPENDENCY_COLLECTION_ENABLED", true) || c.CollectDependencies {
+		depPayload := Dependencies{[]Dependency{}}
+		deps, ok := debug.ReadBuildInfo()
+		if ok {
+			for _, dep := range deps.Deps {
+				depPayload.Dependencies = append(depPayload.Dependencies,
+					Dependency{
+						Name:    dep.Path,
+						Version: dep.Version,
+					},
+				)
+			}
 		}
+		dep := c.newRequest(RequestTypeDependenciesLoaded)
+		dep.Body.Payload = depPayload
+		c.scheduleSubmit(dep)
 	}
-	c.Env = configEnvFallback("DD_ENV", c.Env)
-	c.Version = configEnvFallback("DD_VERSION", c.Version)
 
-	c.APIKey = configEnvFallback("DD_API_KEY", c.APIKey)
-
-	r := c.newRequest(RequestTypeAppStarted)
-	r.Payload = payload
-	c.scheduleSubmit(r)
 	c.flush()
 
-	if c.SubmissionInterval == 0 {
-		c.SubmissionInterval = 60 * time.Second
-	}
-	c.flushT = time.AfterFunc(c.SubmissionInterval, c.backgroundFlush)
-
-	heartbeat := internal.IntEnv("DD_TELEMETRY_HEARTBEAT_INTERVAL", defaultHeartbeatInterval)
-	if heartbeat < 1 || heartbeat > 3600 {
-		log.Warn("DD_TELEMETRY_HEARTBEAT_INTERVAL=%d not in [1,3600] range, setting to default of %d", heartbeat, defaultHeartbeatInterval)
-		heartbeat = defaultHeartbeatInterval
-	}
-	c.heartbeatInterval = time.Duration(heartbeat) * time.Second
 	c.heartbeatT = time.AfterFunc(c.heartbeatInterval, c.backgroundHeartbeat)
```

Comment:
Should this line be moved to `applyDefaultOps` too?

Reply:
I wanted to keep this line at the end of the function so the timer was started only after the app-started and dependencies event were flushed. 
I also think the heartbeat timer should be re-initialized every time the client starts (not completely clear on which conditions will cause the same client to start multiple times, though), and `applyDefaultOps` mainly populates unset client fields. I will rename `applyDefaultOps` to `applyFallbackOps` to make this more clear

Reply:
Update on this: I will actually remove all DD_INSTRUMENTATION_TELEMETRY... env variable related configuration from `applyFallback/DefaultOps` and into the Start function to ensure the telemetry client always starts with the most updated env var values.